pkg/blob/controllerserver.go

```diff
@@ -768,18 +791,62 @@ func (d *Driver) copyBlobContainer(_ context.Context, req *csi.CreateVolumeReque
 }
 
 // copyVolume copies a volume form volume or snapshot, snapshot is not supported now
-func (d *Driver) copyVolume(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName, storageEndpointSuffix string) error {
+func (d *Driver) copyVolume(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName, storageEndpointSuffix, accountSasToken string) error {
 	vs := req.VolumeContentSource
 	switch vs.Type.(type) {
 	case *csi.VolumeContentSource_Snapshot:
 		return status.Errorf(codes.InvalidArgument, "copy volume from volumeSnapshot is not supported")
 	case *csi.VolumeContentSource_Volume:
-		return d.copyBlobContainer(ctx, req, accountKey, dstContainerName, storageEndpointSuffix)
+		return d.copyBlobContainer(ctx, req, accountKey, dstContainerName, storageEndpointSuffix, accountSasToken)
 	default:
 		return status.Errorf(codes.InvalidArgument, "%v is not a proper volume source", vs)
 	}
 }
 
+func (d *Driver) authorizeAzcopyBySecurityPrincipal() error {
+	if len(d.cloud.Config.AADClientSecret) > 0 {
+		klog.V(2).Infof("use service principal to authorize azcopy")
+		if err := os.Setenv(azcopyAutoLoginType, "SPN"); err != nil {
+			return err
+		}
+		if d.cloud.Config.AADClientID == "" || d.cloud.Config.AADClientSecret == "" {
+			return fmt.Errorf("AADClientID and AADClientSecret must be set when use service principal")
+		}
+		if err := os.Setenv(azcopySPAApplicationID, d.cloud.Config.AADClientID); err != nil {
+			return err
+		}
+		if err := os.Setenv(azcopySPAClientSecret, d.cloud.Config.AADClientSecret); err != nil {
+			return err
+		}
+		klog.V(2).Infof("set AZCOPY_AUTO_LOGIN_TYPE=SPN, AZCOPY_SPA_APPLICATION_ID and AZCOPY_SPA_CLIENT_SECRET successfully")
+		if d.cloud.Config.TenantID != "" {
+			if err := os.Setenv(azcopyTenantID, d.cloud.Config.TenantID); err != nil {
+				return err
+			}
+			klog.V(2).Infof("set AZCOPY_AUTO_LOGIN_TYPE=SPN, AZCOPY_SPA_APPLICATION_ID, AZCOPY_SPA_CLIENT_SECRET and AZCOPY_TENANT_ID successfully")
+		}
+		return nil
+	}
+	if d.cloud.Config.AzureAuthConfig.UseManagedIdentityExtension {
+		klog.V(2).Infof("use managed identity to authorize azcopy")
```

Comment:
remove this log since L837, L844 already shows which identity this driver is using now